jlm/llvm/ir/operators/Load.hpp

```diff
@@ -575,6 +577,134 @@ class LoadNonVolatileNode final : public LoadNode
   }
 };
 
+/**
+ * \brief Swaps a memory state merge operation and a load operation.
+ *
+ * sx1 = MemStateMerge si1 ... siM
+ * v sl1 = load_op a sx1
+ * =>
+ * v sl1 ... slM = load_op a si1 ... siM
+ * sx1 = MemStateMerge sl1 ... slM
+ *
+ * FIXME: The reduction can be generalized: A load node can have multiple operands from different
+ * merge nodes.
+ *
+ * @return If the normalization could be applied, then the results of the load operation after
+ * the transformation. Otherwise, std::nullopt.
+ */
+std::optional<std::vector<rvsdg::output *>>
+NormalizeLoadMux(
+    const LoadNonVolatileOperation & operation,
+    const std::vector<rvsdg::output *> & operands);
+
+/**
+ * \brief If the producer of a load's address is an alloca operation, then we can remove all
+ * state edges originating from other alloca operations.
+ *
+ * a1 s1 = alloca_op ...
+ * a2 s2 = alloca_op ...
+ * s3 = mux_op s1
+ * v sl1 sl2 sl3 = load_op a1 s1 s2 s3
+ * =>
+ * ...
+ * v sl1 sl3 = load_op a1 s1 s3
+ *
+ * @param operation The load operation on which the transformation is performed.
+ * @param operands The operands of the load node.
+ *
+ * @return If the normalization could be applied, then the results of the load operation after
+ * the transformation. Otherwise, std::nullopt.
+ */
+std::optional<std::vector<rvsdg::output *>>
+NormalizeLoadAlloca(
+    const LoadNonVolatileOperation & operation,
+    const std::vector<rvsdg::output *> & operands);
+
+/**
+ * \brief Forwards the value from a store operation.
+ *
+ * s2 = store_op a v1 s1
+ * v2 s3 = load_op a s2
+ * ... = any_op v2
+ * =>
+ * s2 = store_op a v1 s1
+ * ... = any_op v1
+ *
+ * @param operation The load operation on which the transformation is performed.
+ * @param operands The operands of the load node.
+ *
+ * @return If the normalization could be applied, then the results of the load operation after
+ * the transformation. Otherwise, std::nullopt.
+ */
+std::optional<std::vector<rvsdg::output *>>
+NormalizeLoadStore(
+    const LoadNonVolatileOperation & operation,
+    const std::vector<rvsdg::output *> & operands);
+
+/**
+ * \brief If the producer of a load's address is an alloca operation, then we can remove all
+ * state edges originating from other alloca operations coming through store operations.
+ *
+ * a1 sa1 = alloca_op ...
+ * a2 sa2 = alloca_op ...
+ * ss1 = store_op a1 ... sa1
+ * ss2 = store_op a2 ... sa2
+ * ... = load_op a1 ss1 ss2
+ * =>
+ * ...
+ * ... = load_op a1 ss1
+ *
+ * @param operation The load operation on which the transformation is performed.
+ * @param operands The operands of the load node.
+ *
+ * @return If the normalization could be applied, then the results of the load operation after
+ * the transformation. Otherwise, std::nullopt.
+ */
+std::optional<std::vector<rvsdg::output *>>
+NormalizeLoadStoreState(
```

Comment:
Is the result of this normalization always applicable for `NormalizeLoadStore`?
If so, I guess we could go all the way to doing load elision in this normalization, but there might be cases im not seeing where that is not possible?

Reply:
I have not seen this, but I guess you are right. However, I would not like to also refactor the normalizations now. I want to get the old normalization interface out. That is my highest priority for now. Looking at the individual normalization comes afterwards. It is a very good point that you raise though.